In [2]:
import math
from collections import Counter

In [3]:
def entropy(outcomes):
    outcome_counts = Counter(outcomes)
    total_outcomes = len(outcomes)

    probabilities = [count / total_outcomes for count in outcome_counts.values()]

    entropy_value = -sum(p * math.log2(p) for p in probabilities)

    return {
        "occurrences": dict(outcome_counts),
        "probabilities": dict(zip(outcome_counts.keys(), probabilities)),
        "entropy": entropy_value,
    }

In [4]:
outcomes = [3, 2, 5, 1, 4, 6, 3, 3, 2, 1]

In [5]:
outcomes_entropy = entropy(outcomes)

In [6]:
print(f'{'occurrences':<15} {outcomes_entropy['occurrences']}')
print(f'{'probabilities':<15} {outcomes_entropy['probabilities']}')
print(f'{'Entropy':<15} {outcomes_entropy['entropy']}')

occurrences     {3: 3, 2: 2, 5: 1, 1: 2, 4: 1, 6: 1}
probabilities   {3: 0.3, 2: 0.2, 5: 0.1, 1: 0.2, 4: 0.1, 6: 0.1}
Entropy         2.4464393446710155



The entropy value of approximately 2.446 for the given sequence of outcomes indicates the average amount of uncertainty or information content associated with each outcome in the sequence.

In this context, a higher entropy value suggests that the outcomes are more unpredictable or varied, while a lower entropy value indicates that the outcomes are more predictable or uniform.

Here's what the entropy value tells us in this specific case:

The entropy is measured in bits if using base-2 logarithm (which is commonly used in information theory).
Each outcome in the sequence contributes to this average uncertainty.
The closer the entropy value is to zero, the more predictable the outcomes are. Conversely, the closer it is to its maximum value (which depends on the number of possible outcomes), the more unpredictable the outcomes are.
In this case, the entropy value of approximately 2.446 suggests that there is a moderate amount of uncertainty or variability in the sequence of outcomes.

In [8]:
def count_unique_numbers(nums: list):
    counts = {}
    for num in nums:
        if num in counts:
            counts[num] += 1
        else:
            counts[num] = 1
    return counts

In [9]:
count_unique_numbers(outcomes)

{3: 3, 2: 2, 5: 1, 1: 2, 4: 1, 6: 1}

In [10]:
def count_unique_nums(nums: list):
    counts = {}
    for num in nums:
        counts[num] = counts.get(nums, 0) + 1
    return counts


In [11]:
count_unique_nums(outcomes)

TypeError: unhashable type: 'list'